File: backend/app/modules/documents/validation.py
"""Upload validation. Never trust the declared content type (docs/pipelines/extraction.md)."""
from __future__ import annotations

import hashlib
from dataclasses import dataclass

from app.core.config import settings
from app.core.errors import (
    CorruptDocumentError,
    EncryptedPdfError,
    FileTooLargeError,
    TooManyPagesError,
    UnsupportedFileError,
)

PDF_MAGIC = b"%PDF-"
JPEG_MAGIC = b"\xff\xd8\xff"
PNG_MAGIC = b"\x89PNG\r\n\x1a\n"


@dataclass(frozen=True)
class ValidatedUpload:
    checksum: str
    size: int
    page_count: int
    mime: str = "application/pdf"


def compute_checksum(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def _looks_like_image(data: bytes) -> bool:
    return data.startswith(JPEG_MAGIC) or data.startswith(PNG_MAGIC)


def normalize_upload_to_pdf(uploads: list[tuple[str, bytes]]) -> tuple[bytes, str]:
    """Accept exactly what `validate_pdf` always has (one PDF), or one-or-more JPEG/PNG
    images -- one per page, in the given order -- and return `(pdf_bytes, mime)` ready
    for it. A photographed answer sheet is normally several separate photos rather
    than one PDF; this lets the upload endpoint accept that directly, with no change
    to validation, storage, or extraction, none of which ever see anything but a
    single PDF blob either way.

    A mixed batch (a PDF alongside images, or anything that isn't a PDF/JPEG/PNG) is
    rejected outright rather than guessed at.
    """
    if not uploads:
        raise UnsupportedFileError("No file was uploaded.")
    if len(uploads) == 1 and uploads[0][1].startswith(PDF_MAGIC):
        return uploads[0][1], "application/pdf"

    images: list[bytes] = []
    for filename, data in uploads:
        if not _looks_like_image(data):
            raise UnsupportedFileError(
                "Upload a single PDF, or one or more JPEG/PNG images (one per page).",
                filename=filename,
            )
        images.append(data)

    from app.modules.documents.pdf import images_to_pdf

    try:
        pdf_bytes = images_to_pdf(images)
    except Exception as exc:  # noqa: BLE001 - any decode failure is a corrupt upload
        raise CorruptDocumentError(str(exc)) from exc
    return pdf_bytes, "application/pdf"


def validate_pdf(data: bytes, declared_mime: str | None = None) -> ValidatedUpload:
    if not data:
        raise UnsupportedFileError("The uploaded file is empty.")
    if len(data) > settings.max_upload_bytes:
        raise FileTooLargeError(size=len(data), cap=settings.max_upload_bytes)
    if not data.startswith(PDF_MAGIC):
        raise UnsupportedFileError(
            "The uploaded file is not a PDF.", declared_mime=declared_mime
        )

    import pymupdf

    try:
        document = pymupdf.open(stream=data, filetype="pdf")
    except Exception as exc:  # noqa: BLE001 - any open failure is a corrupt document
        raise CorruptDocumentError(str(exc)) from exc

    try:
        if document.needs_pass:
            raise EncryptedPdfError()
        page_count = document.page_count
        if page_count == 0:
            raise CorruptDocumentError("The PDF contains no pages.")
        if page_count > settings.max_pages:
            raise TooManyPagesError(page_count=page_count, cap=settings.max_pages)
    finally:
        document.close()

    return ValidatedUpload(
        checksum=compute_checksum(data), size=len(data), page_count=page_count
    )

File: apps/web/lib/server/modules/documents/validation.ts
/**
 * Upload validation. Never trust the declared content type.
 * Port of backend/app/modules/documents/validation.py.
 */
import { createHash } from "crypto";
import { settings } from "@/lib/server/config";
import { FileTooLargeError, UnsupportedFileError, CorruptDocumentError } from "@/lib/server/errors";
import { imagesToPdf, inspectPdf } from "./pdf";

const PDF_MAGIC = Buffer.from("%PDF-");
const JPEG_MAGIC = Buffer.from([0xff, 0xd8, 0xff]);
const PNG_MAGIC = Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]);

export interface ValidatedUpload {
  checksum: string;
  size: number;
  pageCount: number;
  mime: string;
}

export function computeChecksum(data: Buffer): string {
  return createHash("sha256").update(data).digest("hex");
}

function looksLikeImage(data: Buffer): boolean {
  return data.subarray(0, 3).equals(JPEG_MAGIC) || data.subarray(0, 8).equals(PNG_MAGIC);
}

/** Accepts one PDF, or one-or-more JPEG/PNG images (one per page, in order), and
 * returns `(pdfBytes, mime)` ready for validatePdf. A mixed batch is rejected
 * outright rather than guessed at. */
export async function normalizeUploadToPdf(
  uploads: Array<{ filename: string; data: Buffer }>
): Promise<{ data: Buffer; mime: string }> {
  if (uploads.length === 0) throw new UnsupportedFileError("No file was uploaded.");
  if (uploads.length === 1 && uploads[0].data.subarray(0, 5).equals(PDF_MAGIC)) {
    return { data: uploads[0].data, mime: "application/pdf" };
  }

  const images: Buffer[] = [];
  for (const { filename, data } of uploads) {
    if (!looksLikeImage(data)) {
      throw new UnsupportedFileError(
        "Upload a single PDF, or one or more JPEG/PNG images (one per page).",
        { filename }
      );
    }
    images.push(data);
  }

  try {
    return { data: await imagesToPdf(images), mime: "application/pdf" };
  } catch (err) {
    if (err instanceof UnsupportedFileError) throw err;
    throw new CorruptDocumentError(String((err as Error)?.message ?? err));
  }
}

export async function validatePdf(data: Buffer): Promise<ValidatedUpload> {
  if (data.length === 0) throw new UnsupportedFileError("The uploaded file is empty.");
  if (data.length > settings.maxUploadBytes) {
    throw new FileTooLargeError("The uploaded file exceeds the size cap.", {
      size: data.length,
      cap: settings.maxUploadBytes,
    });
  }
  if (!data.subarray(0, 5).equals(PDF_MAGIC)) {
    throw new UnsupportedFileError("The uploaded file is not a PDF.");
  }

  const { pageCount } = await inspectPdf(data);

  return {
    checksum: computeChecksum(data),
    size: data.length,
    pageCount,
    mime: "application/pdf",
  };
}
